core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt

```diff
@@ -81,13 +81,13 @@ class SharedPreferenceUtil @Inject constructor(context: Context?) {
       return storage
     }
 
-  private fun defaultStorage(): String {
-    val externalFilesDir = ContextCompat.getExternalFilesDirs(instance, null)[0]
+  private fun defaultStorage(): String =
+    /*val externalFilesDir = ContextCompat.getExternalFilesDirs(instance, null)[0]
     return if (externalFilesDir != null)
       externalFilesDir.path
     else
-      instance.filesDir.path // workaround for emulators
-  }
+      instance.filesDir.path */ // workaround for emulators
+    ((getExternalFilesDirs(instance, null)[0]))?.let { it.path ?: instance.filesDir.path }!!
```

Comment:
I'm not damn sure about this as i'm not a pro in optimized kotlin usage, that's why i kept in a separate commit so that it may become easier to me keep track of mistakes and learn from them

Reply:
So as a general rule try to avoid `!!`, every time you type it you are saying "My app can crash here". Sometimes it is okay but that is rare, the language designers made it look alarming for a reason.

`((getExternalFilesDirs(instance, null)[0]))`  no need for these brackets

`?.let { it.path ?: instance.filesDir.path }!!` so what you want to be saying here is
`?.let { it. path } ?: instance.filesDir.path`

the first version is
```
if externalFileDir != null { externalFileDir.path or filesDir.path if that is null }
else crash
```

the second version is

```
val dir = if(externalFileDir !=null) {externalFileDir.path} else {null}
if(dir !=null) {dir} else{ filesDir.path}
```

Some people find the [kotlin koans](https://kotlinlang.org/docs/tutorials/koans.html) quite helpful in levelling up their understanding of kotlin features, it does have quite a few.

Reply:
> `?.let { it.path ?: instance.filesDir.path }!!`


Shoot! I had a camouflague for those brackets.